detecting IPSEC SPI unchanged

smagnani96 opened the pull request:
Adding logic (and unit test) to detect unchanged SPI while creating a new IPSEC key.

Fixes: #14873

@pchaigno @julianwiedmann :smiley: 

```release-note
In case of an IPsec key rotation, error if the user forgot to increment the SPI per the documentation.
```

smagnani96 committed:
detecting IPSEC SPI unchanged

This commit introduce a control during the `LoadIPSecKeys` to return
an error when providing a new key (e.g., during rotation) without
incrementing the SPI. In this fix, such case is intercepted and logged
with an error, while preserving status and policies referring to the
latest/previous valid key by not inserting the key in `ipSecKeysRemovalTime`.
Moreover, this commit adjusts the SPI fields in the test data so that they
can preserve their expected behaviour. Finally, a new test is introduced
to verify that `LoadIPSecKeys` effectively returns an error when the SPI
does not while loading a new IPSec key.

This "bug" has a user-visible impact during the IPSec key-rotation:
if performing the key rotation without changing the SPI, an error message
is logged. In this case, the oldKey is not marked to be removed and
is still in use.

Expected error in the daemons logs while reproducing after the fix:
`invalid SPI: changing IPSec keys requires incrementing the key id`.

Fixes: #14873

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

rgo3 commented on pkg/datapath/linux/ipsec/ipsec_linux_test.go:
```diff
@@ -53,6 +53,19 @@ func TestLoadKeysNoFile(t *testing.T) {
+	require.NoError(t, err)
+
+	// Reusing same SPI, must fail.
+	keys = bytes.NewReader(keysAeadDat)
```
Nit: can't we reuse the `keys` variable?

rgo3 commented on pkg/datapath/linux/ipsec/ipsec_linux_test.go:
```diff
@@ -53,6 +53,19 @@ func TestLoadKeysNoFile(t *testing.T) {
 	require.Equal(t, true, os.IsNotExist(err))
 }
 
+func TestLoadKeysSameSPI(t *testing.T) {
```
This test expects behavior I'm a bit worried about: Are we sure that LoadIPSecKeys will never be called on the same key file in practice? Or that the key file has two consecutive entries that reference the same key and SPI? From the original issue I understand that we should only fail if the key changed but SPI didn't. The introduced behavior doesn't exactly match that IIUC.

smagnani96 replied:
I see, thanks for pointing out :smiley: 
The new code should catch that behavior.

rgo3 reviewed (changes requested):
When I was assigned to the issue that's being addressed here, the originally proposed solution involved checking the key and SPI from an existing XFRM state and comparing it to the key and SPI we read in `LoadIPSecKeys`. Overall I prefer the simplicity of the approach here but `LoadIPSecKeys` not behaving the same if it's called on the same key file even if underlying kernel state has not changed worries me a bit.

pchaigno commented on pkg/datapath/linux/ipsec/ipsec_linux_test.go:
```diff
@@ -39,11 +39,14 @@ func setupIPSecSuitePrivileged(tb testing.TB) *slog.Logger {
 
 var (
-	path           = "ipsec_keys_test"
-	keysDat        = []byte("1 hmac(sha256) 0123456789abcdef0123456789abcdef cbc(aes) 0123456789abcdef0123456789abcdef\n1 hmac(sha256) 0123456789abcdef0123456789abcdef cbc(aes) 0123456789abcdef0123456789abcdef\n1 digest_null \"\" cipher_null \"\"\n")
```
Doesn't changing this test case to have a single line also reduces the scope of existing tests?

smagnani96 replied:
Here now I'm leaving the variables as they are, just changing the SPI in each of them to make sure the existing tests don't fail because of this reason. In addition, I'm introducing `keysSameSpiDat` which defines two valid keys but with the same SPI: this is used to cover the new case introduced.

pchaigno commented on pkg/datapath/linux/ipsec/ipsec_linux_test.go:
```diff
@@ -82,6 +84,22 @@ func TestLoadKeys(t *testing.T) {
+	keys := bytes.NewReader(keysSameSpiDat)
+	_, _, err := LoadIPSecKeys(log, keys)
+	require.Error(t, err)
+}
```
Isn't there an implicit dependency between these two test cases? If someone removes the first, or loads a new key in between, it will break the second even though the second won't have been modified. Or did I miss something?

smagnani96 replied:
Right now I opted for introducing and leaving only the `TestLoadKeysSameSPI` test, which tests the actual use case when a new valid key has been provided without increasing the SPI. I think this test should be enough to logically test what is being introduced and addressed with this PR.

pchaigno commented on pkg/datapath/linux/ipsec/ipsec_linux_test.go:
```diff
@@ -39,11 +39,14 @@ func setupIPSecSuitePrivileged(tb testing.TB) *slog.Logger {
+	keyAead256Dat    = []byte("4 rfc4106(gcm(aes)) 44434241343332312423222114131211f4f3f2f144434241343332312423222114131211 128\n")
+	keyInvalidDat    = []byte("1 test abcdefghijklmnopqrstuvwzyzABCDEF test abcdefghijklmnopqrstuvwzyzABCDEF\n")
+	keysSameSpiDat   = []byte("1 hmac(sha256) 0123456789abcdef0123456789abcdef cbc(aes) 0123456789abcdef0123456789abcdef\n1 digest_null \"\" cipher_null \"\"\n")
+	keysDuplicateDat = []byte("6 rfc4106(gcm(aes)) 44434241343332312423222114131211f4f3f2f1 128\n6 rfc4106(gcm(aes)) 44434241343332312423222114131211f4f3f2f1 128\n")
```
Should we keep this exactly the same as `keysSameSpiDat`, just with different SPI and key? If the goal is to check that we mandate a new SPI, then let's test that and nothing else.

pchaigno commented on pkg/datapath/linux/ipsec/ipsec_linux_test.go:
```diff
@@ -82,6 +84,22 @@ func TestLoadKeys(t *testing.T) {
+
+	keys := bytes.NewReader(keysSameSpiDat)
+	_, _, err := LoadIPSecKeys(log, keys)
+	require.Error(t, err)
```
It may be worth checking that we do fail with the expected error here. If the test starts failing with something else, it might hide a regression.

pchaigno commented on pkg/datapath/linux/ipsec/ipsec_linux.go:
```diff
@@ -1157,12 +1157,19 @@ func LoadIPSecKeys(log *slog.Logger, r io.Reader) (int, uint8, error) {
 		}
-		ipSecKeysGlobal[""] = ipSecKey
 
-		ipSecKeysRemovalTime[oldSpi] = time.Now()
```
Moving this line inside the if may change the behavior since it will now be skipped in case the key didn't change. Is there any wider impact?

smagnani96 replied:
I made a different PR discussing it https://github.com/cilium/cilium/pull/34652.
At this point, here we introduce just the check for the same SPI, for which we want to return an error. In this case, I would not insert the `oldKey.Spi`  in the `ipSecKeysRemovalTime` map, since it is not actually removed.
What's your opinion on this?

pchaigno replied:
Yep, it looks good now.

pchaigno reviewed (approved):
I think this is good to go without the first commit.

I'd remove it so we can merge because I suspect the other PR will take a bit longer to get ready.

smagnani96 committed:
ipsec: add `UnsetTestIPSecKey` API for testing purpose

This commit introduces the `UnsetTestIPSecKey` API for testing purpose only.
This function is used to reset the current state of IPSec global variables
during testing. In particular, it helps to prevent loading two IPSec keys
with the same SPI in `node_linux_test`, which would cause the tests to fail
after the previous commit. With this function, it is easy to reset the internal
state of the IPSec-related variables either while executing a test or when
tearing down a test suite. This function is also used in the local
`ipsec_linux_test.go`, to expect a coherent behavior while tearing down
a test suite.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

pchaigno commented on pkg/datapath/linux/node_linux_test.go:
```diff
@@ -94,8 +94,14 @@ var (
 
 	mcastNumIPv4 = []string{"net", "ipv4", "neigh", "default", "mcast_solicit"}
 	mcastNumIPv6 = []string{"net", "ipv6", "neigh", "default", "mcast_solicit"}
+	spi          = uint8(0)
```
Is this really necessary? This sort of global variable in tests can easily create side effects that then end up introducing flakes.

smagnani96 replied:
AFAIU, in these tests we just need an IPSEC key to be present, and we never remove it between these tests.
Possible solutions:

1. Let the tests call the `LoadIPSecKey` with an incremental SPI each time.
2. Let the tests call the `LoadIPSecKey` only once.

Either way, instead of global var, we could wrap the incremental SPI or a boolean variable. Would this be a preferred way? 

```go
var loadOnceIPSecKey func(t *testing.T, log *slog.Logger) = func() func(t *testing.T, log *slog.Logger) {
	var loaded bool
	inner := func(t *testing.T, log *slog.Logger) {
		if loaded {
			return
		}
		keys := bytes.NewReader([]byte("6 rfc4106(gcm(aes)) 44434241343332312423222114131211f4f3f2f1 128\n"))
		_, _, err := ipsec.LoadIPSecKeys(log, keys)
		require.NoError(t, err)
		loaded = true
	}
	return inner
}()
```

pchaigno replied:
Could we clean up the key after each test? That way we avoid state leaking from one test to the next. Then it doesn't matter what each test does because they're all independent.

smagnani96 replied:
I'm scared we don't have a method yet to clean up the key held in `ipSecKeysGlobal`, unless providing a new key with different SPI. Am I missing something? In case, would it be worth adding? 
Otherwise, for the just sake of this test a `sync.Once` could be enough, but I do recognize that what you're suggesting would be a way more elegant solution.

pchaigno replied:
I'm just suggesting to do the cleanup from `setupIPSecSuitePrivileged` after each test instead of once at the end. Am I missing something?

smagnani96 replied:
I think I got what you meant, hopefully :smiley:

rgo3 replied:
IIUC, shouldn't the new API you added, `UnsetTestIPSecKey()`, be called from `setupIPSecSuitePrivileged` as well? Now it's only called in the node test suite.

smagnani96 replied:
> IIUC, shouldn't the new API you added, UnsetTestIPSecKey(), be called from setupIPSecSuitePrivileged as well? Now it's only called in the node test suite.

Since `setupIPSecSuitePrivileged` is within the same package, in theory it wouldn't be necessary to invoke the public API (so far it directly reinitialize the `ipSecKeysGlobal`). What would you suggest?

rgo3 replied:
I don't have a strong opinion on this but since we have a public API that is needed outside the package I don't see why we shouldn't use it inside the package itself to make the cleanup of the key consistent in and outside of the ipsec package?

smagnani96 replied:
I see your point, good to have coherent behavior especially while tearing down test suites. Modified the latest commit, let me know :smiley:

rgo3 replied:
LGTM. I'll remove the discussion label so we can get the PR merged. If Paul still sees something that is missing we can address it in a follow up I presume.

rgo3 reviewed (approved):
Left one comment with regards to cleaning state after individual tests, otherwise LGTM. Sorry for the long delay on my part.
